src/libre/print/dot.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2018 Katherine Flavel
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#include <stdio.h>
+#include <ctype.h>
+
+#include <print/esc.h>
+
+#include "../re_ast.h"
+#include "../re_char_class.h"
+#include "../print.h"
+
+static void
+re_flags_print(FILE *f, enum re_flags fl);
+
+static void
+cc_pp_iter(FILE *f, const struct fsm_options *opt,
+	const void *parent, struct re_char_class_ast *n);
+
+static void
+print_range_endpoint(FILE *f, const struct fsm_options *opt,
+	const struct ast_range_endpoint *r);
+
+static void
+fprintf_count(FILE *f, unsigned count)
+{
+	if (count == AST_COUNT_UNBOUNDED) {
+		fprintf(f, "&#x221e;");
+	} else {
+		fprintf(f, "%u", count);
+	}
+}
+
+static void
+pp_iter(FILE *f, const struct fsm_options *opt,
+	const void *parent, struct ast_expr *n)
+{
+	assert(f != NULL);
+	assert(opt != NULL);
+
+	if (n == NULL) { return; }
```

Comment:
We should soon be able to assert on `n` never being NULL -- the branch with anchoring will also thoroughly switch to using a single interned `EMPTY` node as a sentinel, so a NULL pointer would be an internal error.